lua/feline/generator.lua

```diff
@@ -328,17 +347,109 @@ function M.generate_statusline(winid)
             statusline_type='inactive'
         end
 
-        local statusline = components[statusline_type]
+        local sections = components_table[statusline_type]
+
+        -- Parse all components in the sections and also store the indices of every component so
+        -- both the sections and parsed_sections tables can be accessed later on after components
+        -- are sorted in order of priority
+        -- Also calculate statusline length while doing all of that
+        local parsed_sections = {}
+        local component_indices = {}
+        local statusline_length = 0
+
+        for i, section in ipairs(sections) do
+            parsed_sections[i] = {}
+
+            for j, component in ipairs(section) do
+                local ok, result = pcall(parse_component, component, winid)
+
+                if not ok then
+                    api.nvim_err_writeln(string.format(
+                        "Feline: error while processing component number %d on section %d "..
+                        "of type '%s': %s",
+                        j, i, statusline_type, result
+                    ))
+
+                    result = { str = '', len = 0 }
+                end
+
+                parsed_sections[i][j] = result
+                component_indices[#component_indices+1] = {i, j}
+                statusline_length = statusline_length + result.len
+            end
+        end
+
+        local win_width = api.nvim_win_get_width(winid)
+
+        -- If statusline length is larger than the window width, sort the component indices
+        -- in ascending order of priority of the components they refer to
+        -- Then truncate the components one by one using by their short_provider or hiding them
+        -- entirely until the statusline fits within the window
+        if statusline_length > win_width then
+            table.sort(component_indices, function(a, b)
+                -- Access the original component through the sections table using the indices
+                return (sections[a[1]][a[2]].priority or 0) < (sections[b[1]][b[2]].priority or 0)
```

Comment:
This line is abit too complicated. I would maybe create two variables and compared them to make it more readable.

Reply:
Is it better now?

Reply:
Yeah, better now.